### PR TITLE
Refactor timer items to be a bit more time based

### PR DESCRIPTION
**Pull request by Hirmuolio**
#### Summary
Infrastructure "Refactor timer items to be a bit more time based"

#### Purpose of change

 #65114

Items that activate with timer (grenades and such) use item charges to keep track of their timer. This is bad. Use time for tracking time instead.

#### Describe the solution

Use time_points instead of item charges to determine when item timer should activate.

Since it is time based it won't lose track of time out of reality bubble. They still won't explode when they are outside. But if you set C4 timer to 1 day, leave, and come back 1 day later it will explode at the correct time.

Two ways to define this timer: Adding `"target_timer"` in item transform starts timer when player activates item. `"countdown_interval"` starts timer when item is created.

Removed `"countdown"` item use action. Having multiple ways to start timer was messy. Only one item used it and the functionaly can be replicated with item transform simply by  transforming item into itself.

Since delayed action is no more same as player activation a new `"message"`action was added. Purely cosmetic thing used to tell the player they can't pull the pin on active grenades. Could be removed if it is considered unneeded.

`revert_to` could be defined for tools only. It defines what tool reverts to when to runs out of charges or completes its timer. Now it can be defined to all items and defines what they may revert to after timer.
Items without `revert_to` are deleted at the end of their action. Items that should just stay can revert into themselves.

`"countdown_action"` (probably) works only with actions defined in `iuse_actor.cpp`. The actions defined in `iuse.cpp` interpt the meaning `bool t` differently so they don't work.  
So `"explosion"` and `"transform"` work but `"MOLOTOV"` does not work. Hence those items will still use the old system.

Minor regression: In the old system item with explosion action could emit noise during the explosion countdown. That is not possible anymore as the explosion action only happens at the end of timer instead of happening every turn. So few explosives where this sound was important were not migrated to the new system. Explosives where the sound was not particularly important were migrated and sound was removed.

#### Describe alternatives you've considered

Rewrite `itype::tick` and item processing that follows from there so that player activation and item processing activation are separate things and stop causing problems to each other.

#### Testing

Grenades, molotovs and basic tools seem to work fine.
Active grenades and iridescent carapace plate migrate to the new system on save load.
Item timers stay in save-load cycle.

#### Additional context

Item activation system is one massive sphagetti mess where one thing can be used in two different places for two different purposes and all three systems are intertwined.

### CODE REVIEW

File: data/json/items/tool/explosives.json [hunk 1]
```diff
@@ -108,17 +108,10 @@
     "material": [ "paper", "powder" ],
     "symbol": "*",
     "color": "red",
-    "turns_per_charge": 1,
+    "use_action": { "type": "message", "message": "You've already lit the %s, try throwing it instead.", "name": "Light fuse" },
     "explode_in_fire": true,
     "explosion": { "power": 938 },
```
Comment (Brambor) [inline data/json/items/tool/explosives.json:113]:
```suggestion
    "explosion": { "power": 938 },
    "flags": [ "TRADER_AVOID", "BOMB" ],
```

File: data/json/items/tool/explosives.json [hunk 2]
```diff
@@ -164,17 +156,10 @@
     "material": [ "paper", "powder" ],
     "symbol": "*",
     "color": "red",
-    "turns_per_charge": 1,
+    "use_action": { "type": "message", "message": "You've already lit the %s, try throwing it instead.", "name": "Pull pin" },
     "explode_in_fire": true,
     "explosion": { "power": 938, "shrapnel": { "casing_mass": 1597, "fragment_mass": 0.5 } },
```
Comment (Brambor) [inline data/json/items/tool/explosives.json:161]:
```suggestion
    "explosion": { "power": 938, "shrapnel": { "casing_mass": 1597, "fragment_mass": 0.5 } },
    "flags": [ "TRADER_AVOID", "BOMB" ],
```

**Review by Brambor — COMMENTED** [review]
I think you didn't mean to remove the "flags".

**Comment by Hirmuolio** [comment]
I think this is ready now.

**Comment by andrei8l** [comment]
Needs a test unit

File: src/item.cpp [hunk 3]
```diff
@@ -13231,9 +13239,19 @@ bool item::process_internal( map &here, Character *carrier, const tripoint &pos,
...
+            if( type->countdown_action ) {
+                type->countdown_action.call( carrier ? *carrier : get_avatar(), *this, false, pos );
+            }
+            countdown_point = calendar::turn_max;
+            if( type->revert_to ) {
+                convert( *type->revert_to );
+
+                if( !active ) {
+                    active = needs_processing();
+                }
```
Comment (andrei8l) [inline src/item.cpp:13253]: What's the point of this? `needs_processing()` returns true if `active` is true so this is circular. The other conditions don't need to be mirrored in `active`.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/id-like-to-ask-some-questions-about-the-latest-experimental-version-of-debug/28938/2

**Pull request merged**